oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java

```diff
@@ -0,0 +1,318 @@
+    }
+
+    return exchangeExternalCredentialForAccessToken(stsTokenExchangeRequest.build());
+  }
+
+  @Override
+  public String retrieveSubjectToken() throws IOException {
+    if (identityPoolCredentialSource.credentialSourceType
+        == IdentityPoolCredentialSource.IdentityPoolCredentialSourceType.FILE) {
+      return retrieveSubjectTokenFromCredentialFile();
+    }
+    return getSubjectTokenFromMetadataServer();
+  }
+
+  private String retrieveSubjectTokenFromCredentialFile() throws IOException {
+    String credentialFilePath = identityPoolCredentialSource.credentialLocation;
+    if (!Files.exists(Paths.get(credentialFilePath), LinkOption.NOFOLLOW_LINKS)) {
+      throw new IOException(
+          String.format(
+              "Invalid credential location. The file at %s does not exist.", credentialFilePath));
+    }
+    try {
+      return parseToken(new FileInputStream(new File(credentialFilePath)));
+    } catch (IOException e) {
+      throw new IOException(
+          "Error when attempting to read the subject token from the credential file.", e);
+    }
+  }
+
+  private String parseToken(InputStream inputStream) throws IOException {
+    if (identityPoolCredentialSource.credentialFormatType == CredentialFormatType.TEXT) {
+      BufferedReader reader =
+          new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
+      return CharStreams.toString(reader);
+    }
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    GenericJson fileContents =
+        parser.parseAndClose(inputStream, StandardCharsets.UTF_8, GenericJson.class);
+
+    if (!fileContents.containsKey(identityPoolCredentialSource.subjectTokenFieldName)) {
+      throw new IOException("Invalid subject token field name. No subject token was found.");
+    }
+    return (String) fileContents.get(identityPoolCredentialSource.subjectTokenFieldName);
+  }
+
+  private String getSubjectTokenFromMetadataServer() throws IOException {
+    HttpRequest request =
+        transportFactory
+            .create()
+            .createRequestFactory()
+            .buildGetRequest(new GenericUrl(identityPoolCredentialSource.credentialLocation));
+    request.setParser(new JsonObjectParser(OAuth2Utils.JSON_FACTORY));
+
+    if (identityPoolCredentialSource.hasHeaders()) {
+      HttpHeaders headers = new HttpHeaders();
+      headers.putAll(identityPoolCredentialSource.headers);
+      request.setHeaders(headers);
+    }
+
+    try {
+      HttpResponse response = request.execute();
+      return parseToken(response.getContent());
+    } catch (IOException e) {
+      throw new IOException(
+          String.format("Error getting subject token from metadata server: %s", e.getMessage()), e);
+    }
+  }
+
+  /** Clones the IdentityPoolCredentials with the specified scopes. */
+  @Override
+  public GoogleCredentials createScoped(Collection<String> newScopes) {
```

Comment:
I think this can be declared to return IdentityPoolCredentials thanks to covariant return types